File: cloudflare/src/llm.js
/**
 * OpenRouter / OpenAI-compatible chat → JSON for HelixBench generation.
 */

const QUIZ_SYSTEM = `You are HelixBench, an expert interviewer for pharma computational biology / AI drug discovery.
Generate rigorous, NON-REPETITIVE multiple-choice questions.
Rules:
- Every question must be unique in wording AND concept angle — no paraphrases of each other.
- Prefer BioPython, RDKit, pandas, structure-based design, biologics (BiTE/ADC), docking, pLMs, folding, translational CompBio.
- Exactly 4 choices; exactly one correct; put the correct answer at a RANDOM index (not always 0).
- Explanations must be short and teach something new (1-2 sentences).
- Vary difficulty and include some Python code-reading items when relevant.
Return JSON only:
{"questions":[{"topic":"str","question":"str","code":"optional str","choices":["A","B","C","D"],"answer":0,"explanation":"str"}]}`;

const FLASH_SYSTEM = `You are HelixBench, creating UNIQUE flashcards for pharma CompBio interview prep.
Rules:
- No two cards may share the same prompt or nearly identical answer text.
- Cover distinct subtopics; keep backs dense and practical (2-4 sentences max).
Return JSON only:
{"cards":[{"topic":"str","front":"prompt question","back":"concise answer + why it matters"}]}`;

const MICRO_SYSTEM = `You are HelixBench, generating microlearning PRACTICE for a focused pharma CompBio topic.
Rules:
- Create NEW interview-style questions — do NOT copy lesson sentences verbatim as answer choices.
- No duplicate or near-duplicate questions in the same batch.
- Include Python 'what does this code do' items when useful.
- Exactly 4 choices; correct answer at a random index.
Return JSON only:
{"quiz":[{"topic":"str","question":"str","code":"optional","choices":["A","B","C","D"],"answer":0,"explanation":"str"}],
 "flashcards":[{"topic":"str","front":"str","back":"str"}]}`;

function uuid() {
  return crypto.randomUUID();
}

function extractJson(text) {
  let t = String(text || "").trim();
  if (t.startsWith("```")) {
    t = t.replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/, "");
  }
  try {
    return JSON.parse(t);
  } catch {
    const m = t.match(/(\{[\s\S]*\}|\[[\s\S]*\])/);
    if (!m) throw new Error("Model did not return JSON");
    return JSON.parse(m[1]);
  }
}

function excludeBlock(exclude) {
  if (!Array.isArray(exclude) || !exclude.length) return "";
  const clipped = exclude
    .map((e) => String(e || "").trim().slice(0, 180))
    .filter(Boolean)
    .slice(0, 40);
  if (!clipped.length) return "";
  return (
    "\n\nDO NOT repeat or paraphrase any of these recently used prompts:\n" +
    clipped.map((e) => `- ${e}`).join("\n") +
    "\n"
  );
}

function dedupeQuestions(items, n) {
  const seen = new Set();
  const out = [];
  for (const q of items || []) {
    const key = String(q.question || "")
      .trim()
      .toLowerCase()
      .replace(/\s+/g, " ");
    if (!key || seen.has(key)) continue;
    const choices = q.choices || [];
    if (choices.length !== 4) continue;
    let ans = Number(q.answer ?? 0);
    if (ans < 0 || ans > 3) ans = 0;
    const item = {
      topic: String(q.topic || "CompBio"),
      question: String(q.question || "").trim(),
      choices: choices.map(String),
      answer: ans,
      explanation: String(q.explanation || "").trim(),
      source: "ai",
      id: String(q.id || uuid()),
    };
    if (q.code) item.code = String(q.code);
    if (item.question) {
      seen.add(key);
      out.push(item);
    }
    if (out.length >= n) break;
  }
  return out;
}

function dedupeCards(items, n) {
  const seen = new Set();
  const out = [];
  for (const c of items || []) {
    const front = String(c.front || "").trim();
    const back = String(c.back || "").trim();
    const key = front.toLowerCase().replace(/\s+/g, " ");
    if (!front || !back || seen.has(key)) continue;
    seen.add(key);
    out.push({
      topic: String(c.topic || "CompBio"),
      front,
      back,
      source: "ai",
      id: String(c.id || uuid()),
    });
    if (out.length >= n) break;
  }
  return out;
}

async function chatJson({ system, user, apiKey, baseUrl, model }) {
  if (!apiKey) throw new Error("Missing API key");
  const base = (baseUrl || "https://openrouter.ai/api/v1").replace(/\/$/, "");
  const url = `${base}/chat/completions`;
  const payload = {
    model: model || "openai/gpt-4o-mini",
    temperature: 0.95,
    messages: [
      { role: "system", content: system },
      { role: "user", content: user },
    ],
    response_format: { type: "json_object" },
  };
  const headers = {
    Authorization: `Bearer ${apiKey}`,
    "Content-Type": "application/json",
  };
  if (base.includes("openrouter.ai")) {
    headers["HTTP-Referer"] = "https://github.com/musicofthings/learn-python";
    headers["X-Title"] = "HelixBench";
  }

  let resp = await fetch(url, { method: "POST", headers, body: JSON.stringify(payload) });
  if (resp.status === 400) {
    const errText = await resp.text();
    if (errText.includes("response_format")) {
      delete payload.response_format;
      resp = await fetch(url, { method: "POST", headers, body: JSON.stringify(payload) });
    } else {
      throw new Error(`LLM HTTP 400: ${errText.slice(0, 500)}`);
    }
  }
  if (!resp.ok) {
    const errText = await resp.text();
    throw new Error(`LLM HTTP ${resp.status}: ${errText.slice(0, 500)}`);
  }
  const data = await resp.json();
  const content = data?.choices?.[0]?.message?.content;
  if (!content) throw new Error("Unexpected LLM response");
  return extractJson(content);
}

export async function generateQuizAi({ domain, focus, n, apiKey, baseUrl, model, exclude }) {
  const nonce = uuid().slice(0, 8);
  const user =
    `Domain/category: ${domain}\nFocus: ${focus}\nBatch id: ${nonce}\n` +
    `Generate ${n} brand-new unique interview MCQs. Vary difficulty, topics, and which choice index is correct.\n` +
    `At least ${Math.max(1, Math.floor(n / 3))} items should include a short Python code snippet when relevant.` +
    excludeBlock(exclude);
  const data = await chatJson({ system: QUIZ_SYSTEM, user, apiKey, baseUrl, model });
  const questions = Array.isArray(data) ? data : data.questions;
  if (!Array.isArray(questions) || !questions.length) throw new Error("No questions in model output");
  const cleaned = dedupeQuestions(questions, n);
  if (cleaned.length < Math.max(3, Math.floor(n / 2))) throw new Error("Model returned too few unique questions");
  return cleaned;
}

export async function generateFlashAi({ domain, focus, n, apiKey, baseUrl, model, exclude }) {
  const nonce = uuid().slice(0, 8);
  const user =
    `Domain: ${domain}\nFocus: ${focus}\nBatch id: ${nonce}\n` +
    `Generate ${n} unique flashcards for interview prep.` +
    excludeBlock(exclude);
  const data = await chatJson({ system: FLASH_SYSTEM, user, apiKey, baseUrl, model });
  const cards = Array.isArray(data) ? data : data.cards;
  if (!Array.isArray(cards) || !cards.length) throw new Error("No cards in model output");
  const cleaned = dedupeCards(cards, n);
  if (cleaned.length < Math.max(3, Math.floor(n / 2))) throw new Error("Model returned too few unique cards");
  return cleaned;
}

export async function generateMicroAi({ topicName, quizFocus, nQuiz, nCards, apiKey, baseUrl, model, exclude }) {
  const nonce = uuid().slice(0, 8);
  const user =
    `Microlearning topic: ${topicName}\nFocus areas: ${quizFocus}\nBatch id: ${nonce}\n` +
    `Generate ${nQuiz} unique quiz questions and ${nCards} unique flashcards.\n` +
    `Do not reuse lesson wording as multiple-choice options.` +
    excludeBlock(exclude);
  const data = await chatJson({ system: MICRO_SYSTEM, user, apiKey, baseUrl, model });
  if (!data || typeof data !== "object") throw new Error("Expected JSON object for micro practice");
  const quiz = dedupeQuestions(data.quiz || [], nQuiz);
  const flashcards = dedupeCards(data.flashcards || [], nCards);
  if (!quiz.length && !flashcards.length) throw new Error("Empty micro practice from model");
  return { quiz, flashcards };
}

File: server/llm.py
"""OpenAI-compatible LLM client for dynamic Q&A generation."""

from __future__ import annotations

import json
import os
import re
from typing import Any

import httpx

DEFAULT_BASE_URL = os.getenv("LLM_BASE_URL", "https://api.openai.com/v1")
DEFAULT_MODEL = os.getenv("LLM_MODEL", "gpt-4o-mini")


class LLMError(Exception):
    pass


def _extract_json(text: str) -> Any:
    text = text.strip()
    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?\s*", "", text)
        text = re.sub(r"\s*```$", "", text)
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        m = re.search(r"(\{[\s\S]*\}|\[[\s\S]*\])", text)
        if not m:
            raise LLMError("Model did not return JSON")
        return json.loads(m.group(1))


async def chat_json(
    *,
    system: str,
    user: str,
    api_key: str,
    base_url: str | None = None,
    model: str | None = None,
    temperature: float = 0.8,
) -> Any:
    if not api_key:
        raise LLMError("Missing API key")

    url = (base_url or DEFAULT_BASE_URL).rstrip("/") + "/chat/completions"
    payload = {
        "model": model or DEFAULT_MODEL,
        "temperature": temperature,
        "messages": [
            {"role": "system", "content": system},
            {"role": "user", "content": user},
        ],
        "response_format": {"type": "json_object"},
    }

    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
    }

    async with httpx.AsyncClient(timeout=90.0) as client:
        try:
            resp = await client.post(url, headers=headers, json=payload)
        except httpx.HTTPError as exc:
            raise LLMError(f"LLM request failed: {exc}") from exc

    if resp.status_code >= 400:
        # Retry without response_format for providers that lack it
        if resp.status_code == 400 and "response_format" in resp.text:
            payload.pop("response_format", None)
            async with httpx.AsyncClient(timeout=90.0) as client:
                resp = await client.post(url, headers=headers, json=payload)
        if resp.status_code >= 400:
            raise LLMError(f"LLM HTTP {resp.status_code}: {resp.text[:400]}")

    data = resp.json()
    try:
        content = data["choices"][0]["message"]["content"]
    except (KeyError, IndexError, TypeError) as exc:
        raise LLMError(f"Unexpected LLM response: {data!r}") from exc
    return _extract_json(content)


QUIZ_SYSTEM = """You are HelixBench, an expert interviewer for pharma computational biology / AI drug discovery.
Generate rigorous multiple-choice questions. Prefer BioPython, RDKit, pandas, structure-based design, biologics, and translational CompBio — not generic Python trivia.
Return JSON only with shape:
{"questions":[{"topic":"str","question":"str","code":"optional str","choices":["A","B","C","D"],"answer":0,"explanation":"str"}]}
answer is the index of the correct choice (0-3). Exactly 4 choices. Explanations must teach briefly."""


FLASH_SYSTEM = """You are HelixBench, creating flashcards for pharma CompBio interview prep.
Return JSON only:
{"cards":[{"topic":"str","front":"prompt question","back":"concise answer + why it matters"}]}
Keep backs dense and practical."""


MICRO_SYSTEM = """You are HelixBench, generating microlearning practice for a focused pharma CompBio topic.
Return JSON only:
{"quiz":[{"topic":"str","question":"str","code":"optional","choices":["A","B","C","D"],"answer":0,"explanation":"str"}],
 "flashcards":[{"topic":"str","front":"str","back":"str"}]}
Include at least 2 code-reading questions when the topic involves Python/workflows.
Questions must be specific to the requested topic (BiTE, ADC, pLMs, folding, docking, etc.)."""


async def generate_quiz_ai(
    *,
    domain: str,
    focus: str,
    n: int,
    api_key: str,
    base_url: str | None,
    model: str | None,
) -> list[dict]:
    user = (
        f"Domain/category: {domain}\n"
        f"Focus: {focus}\n"
        f"Generate {n} NEW unique interview MCQs. Vary difficulty. Shuffle which choice is correct.\n"
        f"At least {max(1, n // 3)} items should include a short Python code snippet when relevant."
    )
    data = await chat_json(system=QUIZ_SYSTEM, user=user, api_key=api_key, base_url=base_url, model=model)
    questions = data.get("questions") if isinstance(data, dict) else data
    if not isinstance(questions, list) or not questions:
        raise LLMError("No questions in model output")
    cleaned = []
    for q in questions[:n]:
        choices = q.get("choices") or []
        if len(choices) != 4:
            continue
        ans = int(q.get("answer", 0))
        if ans < 0 or ans > 3:
            ans = 0
        item = {
            "topic": str(q.get("topic") or domain),
            "question": str(q.get("question") or "").strip(),
            "choices": [str(c) for c in choices],
            "answer": ans,
            "explanation": str(q.get("explanation") or "").strip(),
            "source": "ai",
        }
        if q.get("code"):
            item["code"] = str(q["code"])
        if item["question"]:
            cleaned.append(item)
    if len(cleaned) < max(3, n // 2):
        raise LLMError("Model returned too few valid questions")
    return cleaned


async def generate_flashcards_ai(
    *,
    domain: str,
    focus: str,
    n: int,
    api_key: str,
    base_url: str | None,
    model: str | None,
) -> list[dict]:
    user = f"Domain: {domain}\nFocus: {focus}\nGenerate {n} flashcards for interview prep."
    data = await chat_json(system=FLASH_SYSTEM, user=user, api_key=api_key, base_url=base_url, model=model)
    cards = data.get("cards") if isinstance(data, dict) else data
    if not isinstance(cards, list) or not cards:
        raise LLMError("No cards in model output")
    cleaned = []
    for c in cards[:n]:
        front = str(c.get("front") or "").strip()
        back = str(c.get("back") or "").strip()
        if not front or not back:
            continue
        cleaned.append(
            {
                "topic": str(c.get("topic") or domain),
                "front": front,
                "back": back,
                "source": "ai",
            }
        )
    if len(cleaned) < max(3, n // 2):
        raise LLMError("Model returned too few valid cards")
    return cleaned


async def generate_micro_ai(
    *,
    topic_name: str,
    quiz_focus: str,
    n_quiz: int,
    n_cards: int,
    api_key: str,
    base_url: str | None,
    model: str | None,
) -> dict:
    user = (
        f"Microlearning topic: {topic_name}\n"
        f"Focus areas: {quiz_focus}\n"
        f"Generate {n_quiz} quiz questions and {n_cards} flashcards.\n"
        f"Include Python 'what does this code do' items when useful."
    )
    data = await chat_json(system=MICRO_SYSTEM, user=user, api_key=api_key, base_url=base_url, model=model)
    if not isinstance(data, dict):
        raise LLMError("Expected JSON object for micro practice")
    quiz = []
    for q in (data.get("quiz") or [])[:n_quiz]:
        choices = q.get("choices") or []
        if len(choices) != 4:
            continue
        ans = int(q.get("answer", 0))
        item = {
            "topic": str(q.get("topic") or topic_name),
            "question": str(q.get("question") or "").strip(),
            "choices": [str(c) for c in choices],
            "answer": ans if 0 <= ans <= 3 else 0,
            "explanation": str(q.get("explanation") or "").strip(),
            "source": "ai",
        }
        if q.get("code"):
            item["code"] = str(q["code"])
        if item["question"]:
            quiz.append(item)
    cards = []
    for c in (data.get("flashcards") or [])[:n_cards]:
        front = str(c.get("front") or "").strip()
        back = str(c.get("back") or "").strip()
        if front and back:
            cards.append(
                {
                    "topic": str(c.get("topic") or topic_name),
                    "front": front,
                    "back": back,
                    "source": "ai",
                }
            )
    if not quiz and not cards:
        raise LLMError("Empty micro practice from model")
    return {"quiz": quiz, "flashcards": cards}
